ocaml/database/redo_log.ml

```diff
@@ -888,6 +888,26 @@ let flush_db_to_redo_log db log =
     write_db_to_fd log ;
   !(log.currently_accessible)
 
+let flush_db_exn db log =
```

Comment:
Might be worth having a small module here:

```ocaml
module Flush = struct
  type t = private Flush of (Db_cache_types.Database.t * redo_log) | No of redo_log

  val no = No

  val flush db log = if db.read_only then No else Flush db log
end
```
then this can be come
```ocaml
let flush_db_exn = function
| No _ ->
  ()
| Flush (db, log) ->
  R.debug "Flushing..."
```
and can be used like
```ocaml
Redo_log.enable Xapi_ha.(Flush.no ha_redo_log) Xapi_globs.gen_metadata_vdi_reason ;
...
let flush = Flush.flush (Context.database_of __context |> Db_ref.get_database) ha_redo_log in
Redo_log.enable flush Xapi_globs.ha_metadata_vdi_reason
```

Reply:
I've pushed some commits implementing this idea, and split the enable functions into 2: `enable_existing` which doesn't take a database as a parameter and accepts both RO and RW redologs. And `enable_and_flush` which takes only RW redologs and flushes. (enable_existing needs to accept RW because the HA redo log starts out RO, to read back old state, and then becomes RW)